source/extensions/quic_listeners/quiche/codec_impl.cc

```diff
@@ -1,18 +1,34 @@
 #include "extensions/quic_listeners/quiche/codec_impl.h"
 
+#include "extensions/quic_listeners/quiche/envoy_quic_server_stream.h"
+
 namespace Envoy {
 namespace Quic {
 
 bool QuicHttpConnectionImplBase::wantsToWrite() { return quic_session_.HasDataToWrite(); }
 
-// TODO(danzh): modify QUIC stack to react based on aggregated bytes across all
-// the streams. And call StreamCallbackHelper::runHighWatermarkCallbacks() for each stream.
-void QuicHttpConnectionImplBase::onUnderlyingConnectionAboveWriteBufferHighWatermark() {
-  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+QuicHttpServerConnectionImpl::QuicHttpServerConnectionImpl(
+    EnvoyQuicServerSession& quic_session, Http::ServerConnectionCallbacks& callbacks)
+    : QuicHttpConnectionImplBase(quic_session), quic_server_session_(quic_session) {
+  quic_session.setHttpConnectionCallbacks(callbacks);
+}
+
+void QuicHttpServerConnectionImpl::onUnderlyingConnectionAboveWriteBufferHighWatermark() {
+  for (auto& it : quic_server_session_.stream_map()) {
+    if (!it.second->is_static()) {
+      ENVOY_LOG(debug, "runHighWatermarkCallbacks on stream {}", it.first);
+      dynamic_cast<EnvoyQuicServerStream*>(it.second.get())->runHighWatermarkCallbacks();
```

Comment:
Why do we need the dynamic cast here and below? Can the stored stream just derive from the right interface?

Reply:
stream_map() is a QUICHE interface, for sure it doesn't know about Envoy stream interface.

Reply:
I see so this map is storing an opaque pointer of some type? Could it eventually just store a common interface that our classes derive from? Can this be a TODO here and below?

Reply:
I'm happy with the TODO there.  Can we have some comment the first time we reference is_static just explaining it's crypto and gQUIC headers?

Reply:
done